update Hercules P32 mapping

Be-ing opened the pull request:
plus miscellaneous bug fixes and code cleanup

depends on
- [x] #1187
- [x] #1211
- [x] #1218
- [x] #1271

Be-ing committed:
Hercules P32: fix rate_temp_up button

Be-ing committed:
Hercules P32: fix deck toggling

Be-ing committed:
Hercules P32: use new beatloop_size and beatjump_size COs

introduced in PR #1187

Be-ing committed:
Hercules P32: code cleanup

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into hercules_p32_mapping_for_2.1

Be-ing committed:
Hercules P32: don't toggle slip mode when releasing slip button for toggling a deck

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into hercules_p32_mapping_for_2.1

Be-ing committed:
Hercules P32: remove defaultLoopSize & defaultBeatJumpSize options

Now that these are handled by Mixxx.

Be-ing committed:
Hercules P32: rearrange samplers

To match new skin layout

Be-ing committed:
Components: support alternate color for looping SamplerButton

Be-ing committed:
Hercules P32: make looping sampler pads purple

Be-ing committed:
Hercules P32: rearrange effect unit assignment buttons

following the redesign of Deere, effect unit assignment buttons are
associated with decks

Be-ing committed:
Hercules P32: use new Components EffectUnit API

Be-ing committed:
Hercules P32: remove option to toggle effect units with decks

now that Components supports shift + effectFocusButton to toggle EffectUnits

Be-ing committed:
Hercules P32: revert to using loop_halve and loop_double

Now that they do not adjust beatloop_size out of sync with the actual
loop size

Be-ing committed:
Hercules P32: swap encoder mappings for right side

I got confused many times because the encoders are labelled
symmetrically but nothing else on the controller is. So make the entire
deck asymmetrical.

Be-ing committed:
Hercules P32: rework loop pad mode and pressing loop encoder

Be-ing committed:
Hercules P32: use new reloop CO for shift + loop encoder press

Be-ing committed:
Hercules P32: fix display of beatjump_size on controller

Be-ing committed:
Hercules P32: replace loopauto_toggle with beatloop_toggle

Be-ing committed:
Hercules P32: use new beatloop_halve/double COs

and remove reloop CO which was removed from PR #1187

Be-ing committed:
Hercules P32: use standard Components SyncButton functionality

Be-ing committed:
Hercules P32: fix assignment of effect units to headphones

Be-ing committed:
Hercules P32: once again revert to loop_halve/double

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into hercules_p32_mapping_for_2.1

Be-ing committed:
Hercules P32: use reloop_toggle on loop encoder press when looping

beatloop_toggle no longer deactivates loops

Be-ing committed:
Hercules P32: use renamed beatloop_activate ControlObject

Be-ing committed:
Hercules P32: use beatjump_1_forward/backward

now that loop_move_forward/backward have been removed

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into hercules_p32_mapping_for_2.1

daschuer commented:
Who is able to review this?

Be-ing commented:
[Wiki page](https://mixxx.org/wiki/doku.php/hercules_p32_dj) has been updated.

Be-ing committed:
Hercules P32: refactor for new Components button types

Be-ing committed:
Hercules P32: remove dedicated quantize button

now that shift + sync toggles quantize

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into hercules_p32_mapping_for_2.1

Be-ing commented:
Ready for merge? It would be nice if Hercules could test all the new features with a master build.

radusuciu commented on res/controllers/Hercules-P32-scripts.js:
```diff
@@ -218,127 +229,164 @@ P32.Deck = function (deckNumbers, channel) {
-    this.alignBeats.send(P32.padColors.blue);
+
+    this.enableEffectUnitButtons = new components.ComponentContainer(); //fii
+    this.enableEffectUnitButtons[1] = new components.EffectAssignmentButton({
```
Perhaps create these in a loop?

Be-ing replied:
A loop would require some math for calculating the second byte of the MIDI messages, which I think is a bit overcomplicated when there are only 4 iterations. Sometimes it's easier to just copy & paste.

radusuciu replied:
Alternative to math:

```javascript
this.enableEffectUnitButtons = _.map([0x40, 0x41, 0x3c, 0x3d], function(midi, index) {
    return new components.EffectAssignmentButton({
        midi: [0x90 + channel, midi],
        effectUnit: index + 1,
        group: this.currentDeck,
        on: P32.padColors.blue,
        off: P32.padColors.red,
    });
});
```

Be-ing replied:
No need to use Lodash for this, standard Array.prototype.map works.

radusuciu commented on res/controllers/Hercules-P32-scripts.js:
```diff
@@ -218,127 +229,164 @@ P32.Deck = function (deckNumbers, channel) {
+            }, true);
+        } else { // right turn
+            engine.setValue(this.currentDeck, 'beatjump_1_forward', 1);
+            engine.beginTimer(200, function () {
```
Why is this timer necessary here?

Be-ing replied:
So the button on screen lights up. Thanks for the reminder about this. I'll factor this out into a common function for common-controller-scripts.js.

radusuciu replied:
But why delayed by 200ms? I trust the the code is necessary actually and I think I've had to do it as well - at least setting something to 1 and then back to 0 immediately without delay.

Would definitely be nice to have factored out, thanks!

Be-ing replied:
The delay is so the skin button lights up briefly. If it is set to 1 then immediately back to 0 there is no response on screen. For this particular case, there are no skin buttons that light up with beatjump_1_forward, so it doesn't really make a difference. But for buttons like eject it does make a difference.

Be-ing committed:
add script.triggerControl to common-controller-scripts.js

Be-ing committed:
Hercules P32: use new script.triggerControl function

Be-ing committed:
Hercules P32: code cleanup

radusuciu commented on res/controllers/Hercules-P32-scripts.js:
```diff
@@ -445,12 +435,7 @@ P32.Deck = function (deckNumbers, channel) {
-                engine.setValue(this.currentDeck, 'eject', 0);
-            }, true);
-        }
+        engine.setValue(this.currentDeck, 'eject', value / 127);
```
No longer `triggerControl`?

Be-ing replied:
No, I realized that wasn't appropriate here because there is a button up event... although, thinking about it again, this implementation requires that the Load button is released before the Shift button, which is the problem I solved for Components Buttons in #1271. I think I'll refactor this to use Components.

radusuciu commented on res/controllers/midi-components-0.0.js:
```diff
@@ -319,30 +319,45 @@
                         engine.setValue(this.group, 'play', 0);
                     } else {
                         engine.setValue(this.group, 'eject', 1);
+                        engine.beginTimer(225, function () {
```
Can use new `triggerControl` here?

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into hercules_p32_mapping_for_2.1

Be-ing committed:
Hercules P32: refactor to use Components for everything

also rearrange the code from top to bottom in order of how the
controller is arranged

Be-ing committed:
Components minor code cleanup

Be-ing committed:
Hercules P32: make asymmetrical mapping of encoders optional

By default, the mapping is mirrored like the encoders' labels

Be-ing committed:
Hercules P32: clamp loop & beatjump size to 64 by default

The LED displays on the controller can show a maximum of 64.

Be-ing commented:
Ready for merge?

Be-ing commented:
Ping. Ready for merge? The mapping in master has been broken for some time now with API changes to the Components library.

Be-ing commented:
Is there any reason this has not been merged?

daschuer commented:
Should I just merge this now, so we can wait for feedback in the beta phase?

Be-ing commented:
Yes. Evidently no one is going to give feedback until this is merged.

daschuer commented:
Ok, thank you for the work.
